torch/distributed/_tensor/ops/math_ops.py

```diff
@@ -227,6 +232,57 @@ def var_reduction_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
     )
 
 
+@dataclass(frozen=True)
+class _NormPartial(_Partial):
+    norm_type: Union[int, float, str] = 2
+
+    def _partition_value(self):
+        raise NotImplementedError()
+
+    def _reduce_shard_value(self):
+        raise NotImplementedError()
+
+    def _reduce_value(
+        self, tensor: torch.Tensor, mesh: DeviceMesh, mesh_dim: int
+    ) -> torch.Tensor:
+        if self.norm_type in (float("inf"), "inf"):
+            return funcol.all_reduce(tensor, reduceOp="max", group=(mesh, mesh_dim))
+        elif self.norm_type in (float("-inf"), "-inf"):
+            return funcol.all_reduce(tensor, reduceOp="min", group=(mesh, mesh_dim))
+        elif self.norm_type == 0:
+            raise NotImplementedError()
+        elif isinstance(self.norm_type, (int, float)):
+            reduced_tensor = funcol.all_reduce(
+                tensor**self.norm_type, reduceOp="sum", group=(mesh, mesh_dim)
+            )
+            if self.norm_type != 1:
+                reduced_tensor **= 1.0 / self.norm_type
```

Comment:
This means, suppose we have local shards (a -> rank 0, b -> rank 1), then a global norm would be:
global_norm = sum(abs([a b]) ^ {ord}) ^ {1/ord}
each shard norm:
norm_a = sum(abs(a)^ {ord}) ^ {1/ord}
norm_b = sum(abs(b)^ {ord}) ^ {1/ord}

The math we need to figure out is the relation between global_norm and norm_a/b

I wondering if performing `reduced_tensor **= 1.0 / self.norm_type` is the right math 🤔

Reply:
I think it should be `** self.norm_type` -> `all_reduce(sum)` -> `** 1.0 / self.norm_type` (see the other comment).